src/controllerclientimpl.cpp

```diff
@@ -1878,39 +1838,11 @@ void ControllerClientImpl::_UploadFileToControllerViaForm(std::istream& inputStr
     }
 }
 
-void ControllerClientImpl::_UploadDataToController(const std::vector<unsigned char>& vdata, const std::string& desturi)
+void ControllerClientImpl::_UploadDataToController(const void* data, size_t size, const std::string& desturi)
 {
-    curl_off_t filesize = vdata.size();
-
-    // tell it to "upload" to the URL
-    CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_UPLOAD, 0L, 1L);
-    _buffer.clear();
-    _buffer.str("");
-    CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_WRITEFUNCTION, NULL, _WriteStringStreamCallback);
-    CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_WRITEDATA, NULL, &_buffer);
-    CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_HTTPGET, 0L, 0L);
-    CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_HTTPHEADER, NULL, _httpheadersjson);
-    CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_URL, NULL, desturi.c_str());
-    std::pair<std::vector<unsigned char>::const_iterator, size_t> streamdata;
-    streamdata.first = vdata.begin();
-    streamdata.second = vdata.size();
-    CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_INFILESIZE_LARGE, -1, filesize);
-    CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_READFUNCTION, NULL, _ReadInMemoryUploadCallback);
-    CURL_OPTION_SAVE_SETTER(_curl, CURLOPT_READDATA, NULL, &streamdata);
-
-    CURL_PERFORM(_curl);
-    long http_code = 0;
-    CURL_INFO_GETTER(_curl, CURLINFO_RESPONSE_CODE, &http_code);
-
-    // 204 is when it overwrites the file?
-    if( http_code != 201 && http_code != 204 ) {
-        if( http_code == 400 ) {
-            throw MUJIN_EXCEPTION_FORMAT("upload of to failed with HTTP status %s, perhaps file exists already?", desturi%http_code, MEC_HTTPServer);
-        }
-        else {
-            throw MUJIN_EXCEPTION_FORMAT("upload of to failed with HTTP status %s", desturi%http_code, MEC_HTTPServer);
-        }
-    }
+    const std::string filenameoncontroller = desturi.substr(_basewebdavuri.size());
+    std::istrstream stream(reinterpret_cast<const char*>(data), size);
```

Comment:
@ziyan are you sure this is fast?

Reply:
Changed to use the buffer pointer directly with curl.